graph.go

```diff
@@ -4,193 +4,166 @@
-// for the function argument. If HeuristicCost is not intended to be used, it can be implemented as
-// the null heuristic (always returns 0).
-type HeuristicCoster interface {
-	// HeuristicCost returns a heuristic cost between any two nodes.
-	HeuristicCost(n1, n2 Node) float64
+// Weighter wraps the Weight method.
+type Weighter interface {
+	// Weight returns the edge weight for the parameter,
+	Weight(Edge) float64
 }
 
-// A Mutable is a graph that can have arbitrary nodes and edges added or removed.
-//
-// Anything implementing Mutable is required to store the actual argument. So if AddNode(myNode) is
-// called and later a user calls on the graph graph.NodeList(), the node added by AddNode must be
-// an the exact node, not a new node with the same ID.
-//
-// In any case where conflict is possible (e.g. adding two nodes with the same ID), the later
-// call always supercedes the earlier one.
-//
-// Functions will generally expect one of MutableGraph or MutableDirectedGraph and not Mutable
-// itself. That said, any function that takes Mutable[x], the destination mutable should
-// always be a different graph than the source.
+// Mutable wraps generalized graph alteration methods.
 type Mutable interface {
-	// NewNode returns a node with a unique arbitrary ID.
-	NewNode() Node
+	// NewNodeID returns a new unique arbitrary ID.
+	NewNodeID() int
 
 	// Adds a node to the graph. AddNode panics if
 	// the added node ID matches an existing node ID.
 	AddNode(Node)
 
-	// RemoveNode removes a node from the graph, as well as any edges
-	// attached to it. If no such node exists, this is a no-op, not an error.
+	// RemoveNode removes a node from the graph, as
+	// well as any edges attached to it. If the node
+	// is not in the graph it is a no-op.
 	RemoveNode(Node)
-}
 
-// MutableGraph is an interface ensuring the implementation of the ability to construct
-// an arbitrary undirected graph. It is very important to note that any implementation
-// of MutableGraph absolutely cannot safely implement the DirectedGraph interface.
-//
-// A MutableGraph is required to store any Edge argument in the same way Mutable must
-// store a Node argument -- any retrieval call is required to return the exact supplied edge.
-// This is what makes it incompatible with DirectedGraph.
-//
-// The reasoning is this: if you call AddUndirectedEdge(Edge{head,tail}); you are required
-// to return the exact edge passed in when a retrieval method (EdgeTo/EdgeBetween) is called.
-// If I call EdgeTo(tail,head), this means that since the edge exists, and was added as
-// Edge{head,tail} this function MUST return Edge{head,tail}. However, EdgeTo requires this
-// be returned as Edge{tail,head}. Thus there's a conflict that cannot be resolved between the
-// two interface requirements.
-type MutableGraph interface {
-	CostGraph
-	Mutable
+	// SetEdge adds an edge from one node to another.
```

Comment:
The interaction between this and Weighter is not clear. If the graph is not a Weighter, does the cost have to be zero?

I don't have much experience implementing these algorithms (and none in an OO language), but is this really the right way to deal with Edge weights? My gut reaction would be that the weight would be the property of an Edge. I can see the benefit in this implementation, but I just want to make sure it had been considered the other way.